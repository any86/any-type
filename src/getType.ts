import JSON5 from 'json5'

/**
 * 获取json类型
 * @param jsonString json字符串
 * @param space 自定义缩进
 * @returns typescript类型
 */
export default function (jsonString: string, space = '    ') {
  const json = JSON5.parse(jsonString);
  const typeArray: string[] = [`type ResponseData=`];

  /**
   * 遍历节点
   * @param node 节点
   * @param propsName 节点名称 
   * @param level 当前层级, 用来控制缩进
   */
  function walk(node: object, propsName?: string, level = 0) {
    const indent = space.repeat(level);
    if (isObject(node)) {
      if (void 0 === propsName) {
        typeArray.push(`${indent}{`);
      } else {
        typeArray.push(`${indent}${propsName}:{`);
      }
      for (let key in node) {
        walk(node[key], key, level + 1);
      }
      typeArray.push(`${indent}};`);
    } else if (isArray(node)) {
      // wip
      walk(node[0], propsName, level);
      const last = typeArray.pop();
      if (last) {
        typeArray.push(last.replace(';', '[];'));
      }
      // for(let childNode of node){
      //     walk(childNode, propsName, level+1);
      // }
    } else {
      if (void 0 === propsName) {
        typeArray.push(`${indent}${typeof node}`);
      } else {
        typeArray.push(`${indent}${propsName}:${typeof node};`);
      }
    }
  }
  walk(json);
  return (typeArray.join('\r\n'));
}


function isObject(node: unknown): node is Record<string | number | symbol, any> {
  return '[object Object]' === Object.prototype.toString.call(node);
}

function isArray(node: unknown): node is unknown[] {
  return '[object Array]' === Object.prototype.toString.call(node);
}
