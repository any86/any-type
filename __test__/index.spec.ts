import getType from "../src/getType";
test(`验证数组`, () => {
  const input = `[
        {
            "userId": "97",
            "age": 52.487349189738985,
            "sex": 1,
            "nickName": "易芳",
            "userName": "hpk"
        },
        {
            "userId": "60",
            "age": 60.04573151125732,
            "sex": 0,
            "nickName": "徐娜",
            "userName": "tziihyji"
        }
      ]`;

  expect(getType(input)).toMatchSnapshot(`type ResponseData= {
        userId:string;
        age:number;
        sex:number;
        nickName:string;
        userName:string;
    }[];`);
});



// test.skip(`验证对象`, () => {
//     const input = `{
//         "totalSales": {
//             "total": "288769",
//             "weekCompared": "39",
//             "dayCompared": "-44",
//             "dayTotal": "7151"
//         },
//         "order": {
//             "total": "32829",
//             "paid": "46025",
//             "dayTotal": "6442"
//         },
//         "salesRanking": {
//             "list": [
//                 {
//                     "title": "华为手机",
//                     "totalSales": "20851"
//                 },
//                 {
//                     "title": "小米手机",
//                     "totalSales": "30001"
//                 },
//                 {
//                     "title": "苹果手机",
//                     "totalSales": "56678"
//                 }
//             ]
//         },
//         "activityResults": {
//             "percent": 44,
//             "totalOrder": "570290"
//         }
//       }`
  
//     expect(getType(input)).toMatchInlineSnapshot(`type ResponseData=
//       {
//           userId:string;
//           age:number;
//           sex:number;
//           nickName:string;
//           userName:string;
//       }[];`);
//   });
  