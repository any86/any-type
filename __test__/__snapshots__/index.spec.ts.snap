// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`验证数组: type ResponseData= {
        u11serId:string;
        age:number;
        sex:number;
        nickName:string;
        userName:string;
    }[]; 1`] = `
"type ResponseData=
{
    userId:string;
    age:number;
    sex:number;
    nickName:string;
    userName:string;
}[];"
`;

exports[`验证数组: type ResponseData= {
        userId:string;
        age:number;
        sex:number;
        nickName:string;
        userName:string;
    }[]; 1`] = `
"type ResponseData=
{
    userId:string;
    age:number;
    sex:number;
    nickName:string;
    userName:string;
}[];"
`;
